pkg/syncer/upsync/upsync_cleanup_reconcile.go

```diff
@@ -0,0 +1,150 @@
+	logger := klog.FromContext(ctx)
+
+	downstreamNamespace := ""
+	if upstreamNamespace != "" {
+		// find downstream namespace through locator index
+		locator := shared.NewNamespaceLocator(upstreamClusterName, c.syncTargetClusterName, c.syncTargetUID, c.syncTargetName, upstreamNamespace)
+		locatorValue, err := json.Marshal(locator)
+		if err != nil {
+			return nil, err
+		}
+		downstreamNamespaces, err := c.listDownstreamNamespacesByLocator(string(locatorValue))
+		if err != nil {
+			return nil, err
+		}
+		if len(downstreamNamespaces) == 1 {
+			namespace := downstreamNamespaces[0]
+			logger.WithValues(DownstreamName, namespace.GetName()).V(4).Info("Found downstream namespace for upstream namespace")
+			downstreamNamespace = namespace.GetName()
+		} else if len(downstreamNamespaces) > 1 {
+			// This should never happen unless there's some namespace collision.
+			var namespacesCollisions []string
+			for _, namespace := range downstreamNamespaces {
+				namespacesCollisions = append(namespacesCollisions, namespace.GetName())
+			}
+			return nil, fmt.Errorf("(namespace collision) found multiple downstream namespaces: %s for upstream namespace %s|%s", strings.Join(namespacesCollisions, ","), upstreamClusterName, upstreamNamespace)
+		} else {
+			logger.V(4).Info("No downstream namespaces found")
+			// Downstream resource namespace not present => delete resource upstream
+			if err := c.deleteUpstreamResource(ctx, gvr, upstreamClusterName, upstreamNamespace, upstreamName, true); err != nil && !k8serror.IsNotFound(err) {
+				return nil, err
+			}
+			return nil, nil
+		}
+	}
+
+	logger = logger.WithValues(DownstreamNamespace, downstreamNamespace)
+	ctx = klog.NewContext(ctx, logger)
+
+	// retrieve downstream object
+
+	downstreamLister, err := c.getDownstreamLister(gvr)
+	if err != nil {
+		return nil, err
+	}
+
+	var downstreamObject runtime.Object
+	if downstreamNamespace != "" {
+		downstreamObject, err = downstreamLister.ByNamespace(downstreamNamespace).Get(upstreamName)
+	} else {
+		downstreamObject, err = downstreamLister.Get(upstreamName)
+	}
+	if k8serror.IsNotFound(err) {
+		// Downstream resource not present => delete resource upstream
+		if err := c.deleteUpstreamResource(ctx, gvr, upstreamClusterName, upstreamNamespace, upstreamName, true); err != nil && !k8serror.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	downstreamResource, ok := downstreamObject.(*unstructured.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("type mismatch of resource object: received %T", downstreamResource)
+	}
+
+	return downstreamResource, nil
+}
+
+func (c *cleanupReconciler) deleteUpstreamResource(ctx context.Context, gvr schema.GroupVersionResource, clusterName logicalcluster.Name, namespace, name string, force bool) error {
```

Comment:
"force" is a sign of bad factoring. It's unclear without deep understanding of the code what "force" means.